conans/model/requires.py

```diff
@@ -8,28 +8,134 @@
 class Requirement:
     """ A user definition of a requires in a conanfile
     """
-    def __init__(self, ref, *, headers=True, libs=True, build=False, run=None, visible=True,
-                 transitive_headers=None, transitive_libs=None, test=False, package_id_mode=None,
-                 force=False, override=False, direct=True):
+    def __init__(self, ref, *, headers=None, libs=None, build=False, run=None, visible=None,
+                 transitive_headers=None, transitive_libs=None, test=None, package_id_mode=None,
+                 force=None, override=None, direct=None):
         # * prevents the usage of more positional parameters, always ref + **kwargs
         # By default this is a generic library requirement
         self.ref = ref
-        self.headers = headers  # This dependent node has headers that must be -I<headers-path>
-        self.libs = libs
-        self.build = build  # This dependent node is a build tool that is executed at build time only
-        self.run = run  # node contains executables, shared libs or data necessary at host run time
-        self.visible = visible  # Even if not libsed or visible, the node is unique, can conflict
-        self.transitive_headers = transitive_headers
-        self.transitive_libs = transitive_libs
-        self.test = test
-        self.package_id_mode = package_id_mode
-        self.force = force
-        self.override = override
-        self.direct = direct
+        self._headers = headers  # This dependent node has headers that must be -I<headers-path>
+        self._libs = libs
+        self._build = build  # This dependent node is a build tool that is executed at build time only
+        self._run = run  # node contains executables, shared libs or data necessary at host run time
+        self._visible = visible  # Even if not libsed or visible, the node is unique, can conflict
+        self._transitive_headers = transitive_headers
+        self._transitive_libs = transitive_libs
+        self._test = test
+        self._package_id_mode = package_id_mode
+        self._force = force
+        self._override = override
+        self._direct = direct
+
+    @staticmethod
+    def _default_if_none(field, default_value):
+        return field if field is not None else default_value
```

Comment:
There might be a subtle corner case here:
- The propagation is done 2 times, first one it checks for diamonds, previously existing nodes in the graph, with the defined ``requires`` traits, but still without been completed by ``pkg_type``. Second one, after the dependency package has been evaluated and exists, the require is adjusted with the ``pkg_type``, completing Nones
- Previously, if a conflict was detected in this second propagation was because of a runtime (trait ``run``) discrepancy, and the graph was reporting a ``GraphError.runtime`` error.
- Now, this would be generalized for other potential ``None`` values, and the error message could be confusing.